File: prodplus-api/src/main/java/br/com/prodplus/services/ProdutoService.java
package br.com.prodplus.services;

import java.util.List;
import java.util.Set;

import javax.persistence.EntityNotFoundException;
import javax.validation.Valid;
import javax.validation.ValidationException;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.domain.Sort.Direction;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;

import br.com.prodplus.models.Material;
import br.com.prodplus.models.Produto;
import br.com.prodplus.models.auxiliares.OrganoNode;
import br.com.prodplus.models.auxiliares.Quantidade;
import br.com.prodplus.repositories.ProdutoRepository;
import br.com.prodplus.utils.OrganoUtils;

/**
 * 
 * @author Marlon Fernando Garcia
 *
 */
@Service
public class ProdutoService {

	@Autowired
	private ProdutoRepository produtoRepository;

	public Produto salvar(@Valid Produto produto) {
		try {
			produto.setDescricao(produto.getDescricao().toUpperCase().trim());
			return this.produtoRepository.save(produto);
		} catch (DataIntegrityViolationException e) {
			throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "erro de validação!",
					e.getCause());
		} catch (ValidationException e) {
			throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "erro de validação!",
					e.getCause());
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

	public Produto atualizar(Integer id, @Valid Produto produto) {
		try {
			return this.produtoRepository.findById(id).map(p -> {
				p.setAtivo(produto.isAtivo());
				p.setDescricao(produto.getDescricao().toUpperCase().trim());
				p.setTipoMetrico(produto.getTipoMetrico());
				p.setValidade(produto.getValidade());
				return this.produtoRepository.save(p);
			}).orElseThrow(() -> new EntityNotFoundException());
		} catch (EntityNotFoundException e) {
			throw new ResponseStatusException(HttpStatus.NOT_FOUND, "produto não localizado!",
					e.getCause());
		} catch (DataIntegrityViolationException e) {
			throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "erro de validação!",
					e.getCause());
		} catch (ValidationException e) {
			throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "erro de validação!",
					e.getCause());
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

	public Produto buscar(Integer id) {
		try {
			return this.produtoRepository.findById(id)
					.orElseThrow(() -> new EntityNotFoundException());
		} catch (EntityNotFoundException e) {
			throw new ResponseStatusException(HttpStatus.NOT_FOUND, "produto não localizado!",
					e.getCause());
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

	public List<Produto> listar(boolean ativos) {
		try {
			return this.produtoRepository.findAllByAtivo(ativos,
					Sort.by(Direction.ASC, "descricao"));
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

	public Page<Produto> listar(boolean ativos, int pagina, int quant) {
		try {
			Pageable pageable = PageRequest.of(pagina, quant, Direction.ASC, "descricao");
			return this.produtoRepository.findAllByAtivo(ativos, pageable);
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

	public Page<Produto> listar(String descricao, boolean ativos, int pagina, int quant) {
		try {
			Pageable pageable = PageRequest.of(pagina, quant, Direction.ASC, "descricao");
			return this.produtoRepository.findAllByAtivoAndDescricaoContainingIgnoreCase(ativos,
					descricao, pageable);
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

	public void ativar(Integer id) {
		try {
			this.produtoRepository.findById(id).map(p -> {
				p.setAtivo(!p.isAtivo());
				return this.salvar(p);
			}).orElseThrow(() -> new EntityNotFoundException());
		} catch (EntityNotFoundException e) {
			throw new ResponseStatusException(HttpStatus.NOT_FOUND, "produto não localizado!",
					e.getCause());
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

	public void excluir(Integer id) {
		try {
			this.produtoRepository.deleteById(id);
		} catch (DataIntegrityViolationException e) {
			throw new ResponseStatusException(HttpStatus.CONFLICT,
					"não é possível excluir o produto!", e.getCause());
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

	public Set<Quantidade<Material>> getMateriais(Integer id) {
		try {
			return this.produtoRepository.findById(id).map(p -> p.getMateriais())
					.orElseThrow(() -> new EntityNotFoundException());
		} catch (EntityNotFoundException e) {
			throw new ResponseStatusException(HttpStatus.NOT_FOUND, "produto não localizado!",
					e.getCause());
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

	public Set<Quantidade<Produto>> getProdutos(Integer id) {
		try {
			return this.produtoRepository.findById(id).map(p -> p.getProdutos())
					.orElseThrow(() -> new EntityNotFoundException());
		} catch (EntityNotFoundException e) {
			throw new ResponseStatusException(HttpStatus.NOT_FOUND, "produto não localizado!",
					e.getCause());
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

	public Produto setMateriais(Integer id, Set<Quantidade<Material>> materiais) {
		try {
			return this.produtoRepository.findById(id).map(p -> {
				p.setMateriais(materiais);
				return this.salvar(p);
			}).orElseThrow(() -> new EntityNotFoundException());
		} catch (EntityNotFoundException e) {
			throw new ResponseStatusException(HttpStatus.NOT_FOUND, "produto não localizado!",
					e.getCause());
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

	public Produto setProdutos(Integer id, Set<Quantidade<Produto>> produtos) {
		try {
			return this.produtoRepository.findById(id).map(p -> {
				p.setProdutos(produtos);
				return this.salvar(p);
			}).orElseThrow(() -> new EntityNotFoundException());
		} catch (EntityNotFoundException e) {
			throw new ResponseStatusException(HttpStatus.NOT_FOUND, "produto não localizado!",
					e.getCause());
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

	public OrganoNode getOrgano(Integer id, int quantidade) {
		try {
			return OrganoUtils.getOrganograma(this.buscar(id), quantidade);
		} catch (Exception e) {
			e.printStackTrace();
			throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR,
					"ocorreu um erro no servidor!", e.getCause());
		}
	}

}

File: prodplus/src/app/services/produto.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { OrganoNode } from '../models/auxiliares/organo-node';
import { Page } from '../models/auxiliares/page';
import { Quantidade } from '../models/auxiliares/quantidade';
import { Material } from '../models/material';
import { Produto } from '../models/produto';

const URL = environment.url + '/produtos';

@Injectable({ providedIn: 'root' })
export class ProdutoService {
  constructor(private http: HttpClient) {}

  salvar(produto: Produto): Observable<Produto> {
    return this.http.post<Produto>(URL, produto);
  }

  atualizar(id: number, produto: Produto): Observable<Produto> {
    return this.http.put<Produto>(`${URL}/${id}`, produto);
  }

  buscar(id: number): Observable<Produto> {
    return this.http.get<Produto>(`${URL}/${id}`);
  }

  listar(ativos: boolean): Observable<Produto[]> {
    return this.http.get<Produto[]>(`${URL}/listar/${ativos}`);
  }

  listarP(
    ativos: boolean,
    pagina: number,
    quant: number
  ): Observable<Page<Produto>> {
    return this.http.get<Page<Produto>>(
      `${URL}/listar/${pagina}/${quant}/${ativos}`
    );
  }

  listarPD(
    descricao: string,
    ativos: boolean,
    pagina: number,
    quant: number
  ): Observable<Page<Produto>> {
    return this.http.get<Page<Produto>>(
      `${URL}/listar/${pagina}/${quant}/${ativos}/${descricao};`
    );
  }

  ativar(id: number): Observable<any> {
    return this.http.delete(`${URL}/ativar/${id}`);
  }

  excluir(id: number): Observable<any> {
    return this.http.delete(`${URL}/${id}`);
  }

  getMateriais(id: number): Observable<Quantidade<Material>[]> {
    return this.http.get<Quantidade<Material>[]>(`${URL}/materiais/${id}`);
  }

  getProdutos(id: number): Observable<Quantidade<Produto>[]> {
    return this.http.get<Quantidade<Produto>[]>(`${URL}/produtos/${id}`);
  }

  setMateriais(
    id: number,
    materiais: Quantidade<Material>[]
  ): Observable<Produto> {
    return this.http.put<Produto>(`${URL}/materiais/${id}`, materiais);
  }

  setProdutos(
    id: number,
    produtos: Quantidade<Produto>[]
  ): Observable<Produto> {
    return this.http.put<Produto>(`${URL}/produtos/${id}`, produtos);
  }

  getOrgano(id: number, quantidade: number): Observable<OrganoNode> {
    return this.http.get<OrganoNode>(`${URL}/organo/${id}/${quantidade}`);
  }
}
